gapis/api/transform2/control_flow.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/command_generator"
+	"github.com/google/gapid/gapis/config"
+)
+
+type ControlFlow struct {
+	transforms       []Transform
+	commandGenerator command_generator.CommandGenerator
```

Comment:
Command Generators are inside transform chains in the original design, they should not be beside them like this.

Reply:
I disagree with this one. I don't see a parent-child relation between them. Rather they are quite independent.

If the command generator lies in transform, when we get an empty slice out of transform we cannot know if it is empty due to the transform result or the command generator cannot produce anymore command.

I strongly suggest to keep them apart.

Reply:
Are you saying composition should only be used when there "is a parent child relationship" between two classes? There are many reasons why the CommandGenerator is inside the TransformChain in the original design. Among them are: A single place for threading and buffering, encapsulation and protection from outside meddling in the wrapped command generator. Preventing bad-faith uses like pulling from multiple command generators and inserting them all into the same transform chain. 

The TransformChain should also have an EOF() or similarly named function to say when the wrapped CommandGenerator has been exhausted.

Please stick to the original design.